OSDOCS-10980: Updating WAF Tutorials (#77808)
openshift/openshift-docs: cloud_experts_tutorials/cloud-experts-using-alb-and-waf.adoc

You can use an AWS Application Load Balancer (ALB) to add a Web Application Firewall (WAF) to your {product-title} (ROSA) workloads. Using an external solution protects ROSA resources from experiencing denial of service due to handling the WAF.

[IMPORTANT]
====

> It is recommended that you use the significantly more flexible xref:../cloud_experts_tutorials/cloud-experts-using-cloudfront-and-waf.adoc#cloud-experts-using-cloudfront-and-waf[CloudFront method] unless you absolutely must use an ALB based solution.

Reviewer: ```suggestion
It is recommended that you use the more flexible xref:../cloud_experts_tutorials/cloud-experts-using-cloudfront-and-waf.adoc#cloud-experts-using-cloudfront-and-waf[CloudFront method] unless you absolutely must use an ALB based solution.
```

====

[id="prerequisites_{context}"]
== Prerequisites

> * A multi-AZ ROSA (HCP or Classic) cluster.

Reviewer: ```suggestion
* A multiple availability zone (AZ) ROSA or ROSA (classic architecture) cluster.
```

Author: I know we're going to be changing the naming, but it hasn't happened across any of the other ones that I published today, so I kept the (HCP or Classic) until we can make all of these changes across the docs.

> * multiple availability zone (AZ) ROSA (HCP or Classic) cluster.

Reviewer 2: Should this be

```suggestion
* Multiple availability zone (AZ) ROSA (HCP or Classic) cluster.
```

+
[NOTE]
====

> AWS Application Load Balancers (ALB) require at least two _public_ subnets across Availability Zones, link:https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#availability-zones[per the AWS documentation]. For this reason, only multi-AZ ROSA clusters can be used with ALBs.

Reviewer: ```suggestion
AWS ALBs require at least two _public_ subnets across AZs, link:https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#availability-zones[per the AWS documentation]. For this reason, only multiple AZ ROSA clusters can be used with ALBs.
```

+
[source,terminal]
----
$ export VPC_ID=<vpc-id> <1>
$ export PUBLIC_SUBNET_IDS=(<space-separated-list-of-ids>) <2>
$ export PRIVATE_SUBNET_IDS=(<space-separated-list-of-ids>) <3>
----
<1> Replace with the VPC ID of the cluster, for example: `export VPC_ID=vpc-04c429b7dbc4680ba`.

> <2> Replace with a space separated list of the private subnet IDs of the cluster, making sure to preserve the `()`. For example: `export PUBLIC_SUBNET_IDS=(subnet-056fd6861ad332ba2 subnet-08ce3b4ec753fe74c subnet-071aa28228664972f)`.

Reviewer 2: ```suggestion
<2> Replace with a space-separated list of the private subnet IDs of the cluster, making sure to preserve the `()`. For example: `export PUBLIC_SUBNET_IDS=(subnet-056fd6861ad332ba2 subnet-08ce3b4ec753fe74c subnet-071aa28228664972f)`.
```

+
. Add a tag to your cluster's VPC with the cluster identifier:
+
[source,terminal]
----
$ aws ec2 create-tags --resources ${VPC_ID} \
  --tags Key=kubernetes.io/cluster/${CLUSTER},Value=shared --region ${REGION}
----
+
. Add a tag to your public subnets:
+
[source,terminal]
----
$ aws ec2 create-tags \
  --resources ${PUBLIC_SUBNET_IDS} \
  --tags Key=kubernetes.io/role/elb,Value='1' \
        Key=kubernetes.io/cluster/${CLUSTER},Value=shared \
  --region ${REGION}
----
+
. Add a tag to your private subnets:
+
[source,terminal]
----
$ aws ec2 create-tags \
  --resources ${PRIVATE_SUBNET_IDS} \
  --tags Key=kubernetes.io/role/internal-elb,Value='1' \
        Key=kubernetes.io/cluster/${CLUSTER},Value=shared \
  --region ${REGION}
----

[id="deploy-aws-load-balancer-operator_{context}"]
== Deploy the AWS Load Balancer Operator

The link:https://github.com/openshift/aws-load-balancer-operator[AWS Load Balancer Operator] is used to used to install, manage and configure an instance of `aws-load-balancer-controller` in a ROSA cluster. To deploy ALBs in ROSA, we need to first deploy the AWS Load Balancer Operator.

> . Create a new project to deploy the AWS Load Balancer Operator into:

Reviewer: ```suggestion
. Create a new project to deploy the AWS Load Balancer Operator into by running the following command:
```

+
[source,terminal]
----
$ oc new-project aws-load-balancer-operator
----
+

> . Create an AWS IAM policy for the AWS Load Balancer Controller if one does not already exist:

Reviewer: ```suggestion
. Create an AWS IAM policy for the AWS Load Balancer Controller if one does not already exist by running the following command:
```

+
[source,terminal]
----
$ ROLE_ARN=$(aws iam create-role --role-name "${CLUSTER}-alb-operator" \
   --assume-role-policy-document "file://${SCRATCH}/trust-policy.json" \
   --query Role.Arn --output text)
----
+

> . Attach the AWS Load Balancer Operator policy to the IAM role we created previously:

Reviewer: ```suggestion
. Attach the AWS Load Balancer Operator policy to the IAM role we created previously by running the following command:
```

> . Attach the AWS Load Balancer Operator policy to the IAM role we created previously by running the following command:

Reviewer 2: ```suggestion
. Attach the AWS Load Balancer Operator policy to the IAM role you created previously by running the following command:
```

</html
----
+
[NOTE]
====

> Activation of the AWS WAF integration can sometimes take a minute or two. If you do not receive a `403 Forbidden` error, please wait a few seconds and try again.

Reviewer: ```suggestion
Activation of the AWS WAF integration can sometimes take several minutes. If you do not receive a `403 Forbidden` error, please wait a few seconds and try again.
```
